tests/unit/test_utils_temp_dir.py

```diff
@@ -0,0 +1,190 @@
+import itertools
+import os
+import stat
+import tempfile
+
+import pytest
+
+from pip._internal.utils.misc import ensure_dir
+from pip._internal.utils.temp_dir import AdjacentTempDirectory, TempDirectory
+
+
+# No need to test symlinked directories on Windows
+@pytest.mark.skipif("sys.platform == 'win32'")
+def test_symlinked_path():
+    with TempDirectory() as tmp_dir:
+        assert os.path.exists(tmp_dir.path)
+
+        alt_tmp_dir = tempfile.mkdtemp(prefix="pip-test-")
+        assert (
+            os.path.dirname(tmp_dir.path) ==
+            os.path.dirname(os.path.realpath(alt_tmp_dir))
+        )
+        # are we on a system where /tmp is a symlink
+        if os.path.realpath(alt_tmp_dir) != os.path.abspath(alt_tmp_dir):
+            assert (
+                os.path.dirname(tmp_dir.path) !=
+                os.path.dirname(alt_tmp_dir)
+            )
+        else:
+            assert (
+                os.path.dirname(tmp_dir.path) ==
+                os.path.dirname(alt_tmp_dir)
+            )
+        os.rmdir(tmp_dir.path)
+        assert not os.path.exists(tmp_dir.path)
+
+
+def test_deletes_readonly_files():
+    def create_file(*args):
+        fpath = os.path.join(*args)
+        ensure_dir(os.path.dirname(fpath))
+        with open(fpath, "w") as f:
+            f.write("Holla!")
+
+    def readonly_file(*args):
+        fpath = os.path.join(*args)
+        os.chmod(fpath, stat.S_IREAD)
+
+    with TempDirectory() as tmp_dir:
+        create_file(tmp_dir.path, "normal-file")
+        create_file(tmp_dir.path, "readonly-file")
+        readonly_file(tmp_dir.path, "readonly-file")
+
+        create_file(tmp_dir.path, "subfolder", "normal-file")
+        create_file(tmp_dir.path, "subfolder", "readonly-file")
+        readonly_file(tmp_dir.path, "subfolder", "readonly-file")
+
+
+def test_path_access_after_context_raises():
+    with TempDirectory() as tmp_dir:
+        path = tmp_dir.path
+
+    with pytest.raises(AssertionError) as e:
+        _ = tmp_dir.path
+
+    assert path in str(e.value)
+
+
+def test_path_access_after_clean_raises():
+    tmp_dir = TempDirectory()
+    path = tmp_dir.path
+    tmp_dir.cleanup()
+
+    with pytest.raises(AssertionError) as e:
+        _ = tmp_dir.path
```

Comment:
```suggestion
        tmp_dir.path  # this access should fail now.
```

This does not get optimized away.